build_tools/cmake/iree_lit_test.cmake

```diff
@@ -0,0 +1,94 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+include(CMakeParseArguments)
+
+# iree_lit_test()
+#
+# Creates a lit test for the specified source file.
+#
+# Mirrors the bzl rule of the same name.
+#
+# Parameters:
+# NAME: Name of the target
+# TEST_FILE: Test file to run with the lit runner.
+# DATA: Additional data dependencies invoked by the test (e.g. binaries
+#   called in the RUN line)
+#
+# TODO(gcmn): allow using alternative driver
+# A driver other than the default iree/tools/run_lit.sh is not currently supported.
+function(iree_lit_test)
+  cmake_parse_arguments(
+    _RULE
+    ""
+    "NAME;TEST_FILE"
+    "DATA"
+    ${ARGN}
+  )
+  if(NOT IREE_BUILD_TESTS)
+    return()
+  endif()
+
+  iree_package_name(_PACKAGE_NAME)
+  set(_NAME "${_PACKAGE_NAME}_${_RULE_NAME}")
+
+  get_filename_component(_TEST_FILE_PATH ${_RULE_TEST_FILE} ABSOLUTE)
+  
+  add_test(
+    NAME ${_NAME}
+    COMMAND ${CMAKE_SOURCE_DIR}/iree/tools/run_lit.sh "${_TEST_FILE_PATH}"
+    WORKING_DIRECTORY "${CMAKE_BINARY_DIR}" # Make sure the lit runner can find all the binaries
+  )
+  set_tests_properties(${_NAME} PROPERTIES DEPENDS _RULE_DATA)
+endfunction()
+
+
+# iree_lit_test_suite()
+#
+# Creates a suite of lit tests for a list of source files.
+#
+# Mirrors the bzl rule of the same name.
+#
+# Parameters:
+# NAME: Name of the target
+# SRCS: List of test files to run with the lit runner. Creates one test per source.
+# DATA: Additional data dependencies invoked by the test (e.g. binaries
+#   called in the RUN line)
+#
+# TODO(gcmn): allow using alternative driver
+# A driver other than the default iree/tools/run_lit.sh is not currently supported.
+function(iree_lit_test_suite)
+  cmake_parse_arguments(
+    _RULE
+    ""
+    "NAME"
+    "SRCS;DATA"
+    ${ARGN}
+  )
+  IF(NOT IREE_BUILD_TESTS)
+    return()
+  endif()
+
+  foreach(_TEST_FILE ${_RULE_SRCS})
+    get_filename_component(_TEST_BASENAME ${_TEST_FILE} NAME)
+    iree_lit_test(
+      NAME
+        "${_RULE_NAME}_${_TEST_BASENAME}_test"
```

Comment:
tl;dr: Wrote a bunch of point-by-point replies and then realize they all amounted to the same thing. Basically I mirrored the behavior I implemented in the bazel rules. I guess we should have all these discussions about both bazel and cmake together. I'm not trying to hide behind the bazel implementation as gospel. After all, I did just write it last week :-D But I was following some pretty standard conventions there, although maybe they're worth revisiting given our dual build system. Unless there's a strong reason for the two to diverge, I'd like to leave them the same.

> `_test` is already included in the `${_RULE_NAME}_${_TEST_BASENAME}`. Thus, you might want to skip appending `_test`?

You mean because it's in the directory name? That's just because we happen to put them there. We could have lit tests living in a directory without test in the name. It's common in bazel to suffix tests with _test because they often live in the same directory as the things they're testing.

> In addition, to you really want to include `lit` in the name of the test? It somehow more a matter of taste than a serious technical question ;D

The standard for bzl macros is to prefix any generated rules with the name assigned to the macro so that there can't be conflicts.

> One further question is if the absolute test name should be determined in iree_lit_test(). This would allow to use iree_lit_test() also independent of iree_lit_test_suite(). In that case iree_lit_test_suite() would just handle splitting the SRCS list.

Are you suggesting iree_lit_test not having a name argument and just taking a source file? In bzl, it's standard for every rule/macro to have a name so it's clear what all targets are called. `iree_lit_test` should be usable outside of `iree_lit_test_suite` though. Do you see a way that it's not?

Reply:
> tl;dr: Wrote a bunch of point-by-point replies and then realize they all amounted to the same thing. Basically I mirrored the behavior I implemented in the bazel rules. I guess we should have all these discussions about both bazel and cmake together. I'm not trying to hide behind the bazel implementation as gospel. After all, I did just write it last week :-D But I was following some pretty standard conventions there, although maybe they're worth revisiting given our dual build system. Unless there's a strong reason for the two to diverge, I'd like to leave them the same.

Thank you for the detailed response. We definitely should avoid any diverge if possible. So tl;dr, please go ahead and merge without further modifications :D

> You mean because it's in the directory name? That's just because we happen to put them there. We could have lit tests living in a directory without test in the name. It's common in bazel to suffix tests with _test because they often live in the same directory as the things they're testing.

In that case, just add the suffix. In CMake tests can have the same name as build targets.

> The standard for bzl macros is to prefix any generated rules with the name assigned to the macro so that there can't be conflicts.

Agreed.

> Are you suggesting iree_lit_test not having a name argument and just taking a source file? In bzl, it's standard for every rule/macro to have a name so it's clear what all targets are called. `iree_lit_test` should be usable outside of `iree_lit_test_suite` though. Do you see a way that it's not?

Well I an alternative would be to just pass the name + a filename extension (if there should be files other than *.mlir). Anyway, it's fine the way you implemented it.